### ISSUE TITLE
proposal: driver refactor

### ISSUE THREAD

**Issue by ehazlett** [docker/machine#554]
This will pull some of the non-provider operations (provisioning, ssh, etc) out of the driver and enable to driver to only manage instance operations.

This will also generate a driver specification to ensure consistency among the drivers.  Right now, each driver does a little bit of everything: provisioning the actual instance on the provider, custom ssh setups, custom Docker provisioning.  This will get every driver on the same level and make it easier to maintain as well as develop new drivers for machine.

Each driver should be responsible for the following provider actions (if supported):

- Create
- Remove
- Start
- Stop
- Kill
- Restart
- Status

Basically, just the provider specific operations.  SSH commands, Docker provisioning and configuration, etc should be refactored out so the driver only does provider operations.

Tasks
- [x] #743 
- [ ] #744

**Comment by andrewwatson** [comment]
what can i do to help with this?

**Comment by ehazlett** [comment]
@andrewwatson commenting / sending PRs on this (https://github.com/docker/machine/pull/550) would help greatly -- this specifically will probably come after that as I think we need to get that fixed first (moving store, etc).  thanks!

**Comment by nathanleclaire** [comment]
@ehazlett Yes, please!

One thought off the top of my head:  Let's please make `Provisioner` an `interface` for easier testing and so that we can eventually bootstrap on non-Debian / b2d hosts.

**Comment by nathanleclaire** [comment]
FYI @ehazlett : I have a big patch bomb in the works that will fix most of these issues (anticipate getting that in today)

**Comment by nathanleclaire** [comment]
(for making the SSH operations more agnostic / pulling provisioning out of drivers / saving sooner after create - I don't think it steps on the toes of #550 _too_ much)

**Comment by ehazlett** [comment]
lol ok -- perhaps next time we can discuss prior to patch bombs :)

**Comment by nathanleclaire** [comment]
Heh, yeah- that's why I'm letting you know now before I make the PR.  It's turned out to be a bit of a bigger change than expected.

**Comment by ehazlett** [comment]
Ok well in that case let's try to focus on getting a solid design in place and then we can work on it.  I think we get sidetracked solving other problems which pulls us in different directions.  What I wanted from this was a general consensus that it was a good direction and feedback.  I realize we don't have a lot of contributors yet so I think it's fine.  What I didn't want was to have multiple implementations.  I don't care who does it just that we are all on the same page.

I would rather have separate proposal PRs for splitting SSH, provisioning, and the store refactor (which I think is where the saving would go).  I am working on a rough design for the driver refactor but if anyone else has ideas please speak up - no one has until now so it has the potential to conflict.

@nathanleclaire please feel free to open a PR.  We can always review and see if it makes sense.

**Comment by nathanleclaire** [comment]
> What I didn't want was to have multiple implementations. 

Yes, I agree - that's why I wanted to tip you off - next time I will make sure to let you know what I am planning before beginning work.

> I am working on a rough design for the driver refactor but if anyone else has ideas please speak up - no one has until now so it has the potential to conflict.

Agreed again - I will open a PR ASAP

**Comment by ehazlett** [comment]
cool thx

**Comment by nathanleclaire** [comment]
This is working well for me so far.  Virtualbox, DigitalOcean, and EC2 have all worked well without issues except for the lack of SSH username in existing VM like mentioned in IRC.

**Comment by sthulb** [comment]
Same for me. I've tried EC2 + VMware

**Comment by ehazlett** [comment]
@nathanleclaire i've addressed the SSH username issue here https://github.com/docker/machine/pull/756#issuecomment-78416155  thx!
